actors/builtin/miner/miner_actor.go

```diff
@@ -2067,6 +2068,288 @@ func (a Actor) RepayDebt(rt Runtime, _ *abi.EmptyValue) *abi.EmptyValue {
+	// Validate inputs
+
+	// TODO: check how this gets rendered and fix as needed
+	builtin.RequireParam(rt, len(params.Updates) <= ProveReplicaUpdatesMaxSize, "too many updates", len(params.Updates), ProveReplicaUpdatesMaxSize)
+
+	// TODO: Any more quick checks before we jump into it?
+	// see extendsectorexpirations for more checks to plagiarize
+
+	// Kinda sad to be doing (and pay this before a bunch of checks, but oh well?
+	rewRet := requestCurrentEpochBlockReward(rt)
+	powRet := requestCurrentTotalPower(rt)
+
+	store := adt.AsStore(rt)
+	var stReadOnly State
+	rt.StateReadonly(&stReadOnly)
+	info := getMinerInfo(rt, &stReadOnly)
+
+	rt.ValidateImmediateCallerIs(append(info.ControlAddresses, info.Owner, info.Worker)...)
+
+	sectors, err := LoadSectors(store, stReadOnly.Sectors)
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sectors array")
+
+	powerDelta := NewPowerPairZero()
+	pledgeDelta := big.Zero()
+
+	type updateAndSectorInfo struct {
+		update     *ReplicaUpdate
+		sectorInfo *SectorOnChainInfo
+	}
+
+	var sectorsDeals []market.SectorDeals
+	var sectorsDataSpec []*market.SectorDataSpec
+	// Filters out the first wave of invalid updates with quick checks about sector status
+	var prevalidatedUpdates []*updateAndSectorInfo
+	sectorNumbers := bitfield.New()
+	for _, update := range params.Updates {
+		// Bitfied.IsSet() is fast when there are only locally-set values.
+		set, err := sectorNumbers.IsSet(uint64(update.SectorID))
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "error checking sector number")
+		if set {
+			rt.Log(rtt.INFO, "duplicate sector being updated %d, skipping", update.SectorID)
+			continue
+		}
+
+		sectorNumbers.Set(uint64(update.SectorID))
+
+		// TODO: actual length
+		if len(update.Proof) <= 10 {
+			rt.Log(rtt.INFO, "update proof is wrong length (%d), skipping sector %d", len(update.Proof), update.SectorID)
+			continue
+		}
+
+		if len(update.Deals) <= 0 {
+			rt.Log(rtt.INFO, "must have deals to update, skipping sector %d", update.SectorID)
+			continue
+		}
+
+		if uint64(len(update.Deals)) >= SectorDealsMax(info.SectorSize) {
+			rt.Log(rtt.INFO, "more deals than policy allows, skipping sector %d", update.SectorID)
+			continue
+		}
+
+		// TODO Confirm the sector is actually in the stated deadline? This happens later when we try to update partition, but maybe do here first?
+
+		sectorInfo, err := sectors.MustGet(update.SectorID)
+		if err != nil {
+			rt.Log(rtt.INFO, "failed to get sector, skipping sector %d", update.SectorID)
+			continue
+		}
+
+		// TODO: this will change to "was originally dealless, and is currently no active deals"
+		if len(sectorInfo.DealIDs) != 0 {
+			rt.Log(rtt.INFO, "cannot update sector with deals, skipping sector %d", update.SectorID)
+			continue
+		}
+
+		// TODO: confirm sector is healthy (happens implicitly later, but should be done ahead of time)
```

Comment:
I don't know what this means.  Are you planning to make any changes based on this?

Reply:
Oh, this was an open question -- should we confirming that the sector is not faulty at this stage? I'm leaning towards yes, but unsure.

It happens later on when we adjust the partitions, but Idk whether that's sufficient.

Reply:
just realized this is explicitly called necessary in the FIP, so we should do it? https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0019.md#spec

Reply:
We definitely need to make sure it's not terminated, faulty, or unproven before updating.  Whether we do that here or later is a tradeoff as talked about above.

Reply:
Note the current code is safe because this is being done in [ReplaceSectors](https://github.com/filecoin-project/specs-actors/pull/1529/files#diff-d807293f55563b410c2d07719d2f38727016f41c6a17df8635500abc4070f74cR2302) implicitly. But it would be cleaner to explicitly look for termination, faulty and unproven without relying on behavior of ReplaceSectors

The way to do this is to load the partition and check that `partition.ActiveSectors()` contains the sector number in question